packages/server-wallet/src/onchain-service/index.ts

```diff
@@ -0,0 +1,228 @@
+const DEFAULT_MAX_RETRIES = 5;
+const KNOWN_ERRORS = [BAD_NONCE, NO_TX_HASH, UNDERPRICED_REPLACEMENT, INVALID_NONCE];
+
+type OnchainServiceEnvironment = {
+  provider: string | providers.JsonRpcProvider;
+  wallet: string | Wallet; // Private key or wallet to send tx
+  transactionRetries?: number; // Maximum number of times a tx is retried
+};
+
+// This is used instead of the ethers `Transaction` because that type
+// requires the nonce and chain ID to be specified, when sometimes those
+// arguments are not known at the time of creating a transaction.
+type MinimalTransaction = {
+  to: Address;
+  value: string;
+  data: string;
+};
+
+type AssetHolderInformation = {
+  assetId: Address;
+  contractAddress: Address;
+};
+
+export interface OnchainServiceInterface {
+  // NOTE: will return void and throw an error in final implementation
+  registerChannel(
+    channelId: Bytes32,
+    assetHolders: AssetHolderInformation[]
+  ): Promise<string | undefined>;
+  submitTransaction(
+    channelId: Bytes32,
+    tx: MinimalTransaction
+  ): Promise<providers.TransactionResponse>;
+
+  // TODO: remove in v1
+  attachChannelWallet(wallet: ChannelWallet): void;
+}
+
+export class OnchainService implements OnchainServiceInterface {
+  private provider: providers.JsonRpcProvider;
+  private wallet: Wallet;
+  private retries: number;
+
+  // NOTE: V0 has an in-memory store, which will eventually be
+  // shifted to more permanent storage
+  // Holds latest channel event by type (i.e. { "fundsDeposited": {...} })
+  private store: Map<string, object> = new Map();
+
+  // Stores references to all contracts in memory
+  private assetHolders: Map<string, AssetHolderInformation & {contract: Contract}> = new Map();
+
+  // Transaction queue
+  private queue = new PriorityQueue({concurrency: 1});
+
+  // Memory tracking wallet nonce to assist with retries
+  private memoryNonce = 0;
+
+  // TODO: Remove reference to this object, and instead replace with
+  // outbox based communication. Right now used to call correct callbacks
+  private channelWallet: ChannelWallet | undefined = undefined;
+
+  constructor(env: OnchainServiceEnvironment) {
+    this.retries = env.transactionRetries || DEFAULT_MAX_RETRIES;
+    this.provider =
+      typeof env.provider === 'string' ? new providers.JsonRpcProvider(env.provider) : env.provider;
+    this.wallet =
+      typeof env.wallet === 'string'
+        ? new Wallet(env.wallet, this.provider)
+        : (env.wallet as Wallet).connect(this.provider);
+  }
+
+  public attachChannelWallet(wallet: ChannelWallet): void {
+    this.channelWallet = wallet;
+  }
+
+  /**
+   * Adds a channel to watch events and submit transactions for
+   *
+   * @param channelId Unique channel identifier to register chain listeners for
+   * @param assetHolders Onchain addresses of the channel chain context
+   * @returns String if there is an error, undefined if successful
```

Comment:
Does this function ever return a string? Also, this function is not `async` but returns a Promise. Should the function return a Promise?

Reply:
There's a note on the string vs. void in the interface, it was a bit of a hack but I've removed it.

Also, I know right now it's synchronous, but I have a feeling that as soon as we start adding storage calls it will need the asynchronous interface, so I return a `Promise.resolve()` in most places